File: 1.1_The_Elements_of_Programming/Exercise 1.4.cpp
// sicp.exercise.1.4
//-------------------------------------------------------
// * Problem : 
//   Observe that our model of evaluation allows for combinations whose operators are
//   compound expressions. Use this observation to describe the behavior of the following procedure:

#include <iostream>                        // cout, endl, << 
#include <boost/phoenix.hpp>               // arg1, arg2, +, *, >
  
using namespace boost::phoenix::arg_names; // arg1, arg2, +, *, >


auto gt  = arg1 > arg2;
auto add = arg1 + arg2;
auto sub = arg1 - arg2;

//  (define (a-plus-abs-b a b)
//      ((if (> b 0) + -) a b))
auto a_plus_abs_b = 
         if_else(gt (arg2, 0), add, sub);


auto print = [](auto v) {  std::cout << v << std::endl; };

//-------------------------------------------------------
// * Test
//(a-plus-abs-b 10 -4)
//(a-plus-abs-b 10 +4)
int main()
{
  print (a_plus_abs_b (10, -4));
  print (a_plus_abs_b (10, +4));
  
}
//========================================================
// * Output
//--------------------------------------------------------
//14
//14

//-------------------------------------------------------
// * Answer

// a_plus_abs_b                  (10, -4)
// if_else(gt (-4, 0), add, sub) (10, -4)
// if_else(false     , add, sub) (10, -4)
//                          sub  (10, -4)
//                          14

// a_plus_abs_b                  (10, +4)
// if_else(gt (+4, 0), add, sub) (10, +4)
// if_else(true      , add, sub) (10, +4)
//                     add       (10, +4)
//                     14

File: 1.1_The_Elements_of_Programming/Exercise 1.4.hs
-- sicp.exercise.1.4
---------------------------------------------------------
-- * Problem : 
--   Observe that our model of evaluation allows for combinations whose operators are
--   compound expressions. Use this observation to describe the behavior of the following procedure:

--  (define (a-plus-abs-b a b)
--      ((if (> b 0) + -) a b))
a_plus_abs_b a b =
        (if b > 0 then (+) else (-)) a b
    
---------------------------------------------------------
-- * Test
-- (a-plus-abs-b 10 -4)
-- (a-plus-abs-b 10 +4)
main = do
    print $ a_plus_abs_b 10 (-4)
    print $ a_plus_abs_b 10   4
    
-- ========================================================
-- * Output
----------------------------------------------------------
--14
--14

---------------------------------------------------------
-- * Answer

-- a_plus_abs_b                    10 (-4)
-- (if (-4) > 0 then (+) else (-)) 10 (-4)
-- (if False    then (+) else (-)) 10 (-4)
--                            (-)  10 (-4)
--                            14

-- a_plus_abs_b                    10 4
-- (if 4 > 0    then (+) else (-)) 10 4
-- (if True     then (+) else (-)) 10 4
--                   (+)           10 4
--                   14
